File: backend/app/routes/events.py
import uuid

from flask import Blueprint, jsonify, request
from flask_jwt_extended import get_jwt_identity, jwt_required, verify_jwt_in_request

from app.decorators import role_required
from app.extensions import db
from app.models import ClubMember, Event, User
from app.serializers import event_to_dict

bp = Blueprint("events", __name__, url_prefix="/api/events")


def _can_manage_event(user: User, club_id: str) -> bool:
    if not user:
        return False
    if user.role == "admin":
        return True
    if user.role == "club_head" and user.club_id == club_id:
        return True
    return False


@bp.route("", methods=["GET"])
def list_events():
    """
    Role-specific event display:
    - admin: all events
    - club_head: only their club's events
    - student: all events from clubs in which the student is enrolled + approved events
    - guest: approved events only (fallback)
    """
    uid = None
    try:
        verify_jwt_in_request(optional=True)
        uid = get_jwt_identity()
    except Exception:
        pass

    if not uid:
        # Guest: approved events only
        events = Event.query.filter_by(status="approved").all()
        return jsonify([event_to_dict(e) for e in events]), 200

    user = db.session.get(User, uid)
    if not user:
        events = Event.query.filter_by(status="approved").all()
        return jsonify([event_to_dict(e) for e in events]), 200

    if user.role == "admin":
        events = Event.query.order_by(Event.date.desc()).all()
    elif user.role == "club_head":
        events = Event.query.filter_by(club_id=user.club_id).order_by(Event.date.desc()).all()
    elif user.role == "student":
        memberships = ClubMember.query.filter_by(user_id=uid).all()
        club_ids = [m.club_id for m in memberships]
        events = Event.query.filter(
            db.or_(
                Event.status == "approved",
                Event.club_id.in_(club_ids) if club_ids else False
            )
        ).order_by(Event.date.desc()).all()
    else:
        events = Event.query.filter_by(status="approved").all()

    return jsonify([event_to_dict(e) for e in events]), 200


@bp.route("/<eid>", methods=["GET"])
def get_event(eid):
    ev = db.session.get(Event, eid)
    if not ev:
        return jsonify({"message": "Not found"}), 404
    return jsonify(event_to_dict(ev)), 200


@bp.route("", methods=["POST"])
@jwt_required()
@role_required("admin", "club_head")
def create_event():
    uid = get_jwt_identity()
    user = db.session.get(User, uid)
    if not user:
        return jsonify({"message": "Unauthorized"}), 401

    data = request.get_json(silent=True) or {}
    club_id = data.get("clubId") or data.get("club_id")
    if not club_id:
        return jsonify({"message": "clubId required"}), 400
    if not _can_manage_event(user, club_id):
        return jsonify({"message": "Forbidden"}), 403

    title = (data.get("title") or "").strip()
    if not title:
        return jsonify({"message": "Title required"}), 400

    eid = f"event-{uuid.uuid4().hex[:12]}"
    created_by = data.get("createdBy") or data.get("created_by") or uid
    
    event_status = "approved" if user.role == "admin" else "pending"

    ev = Event(
        id=eid,
        title=title,
        description=data.get("description") or "",
        date=data.get("date") or "",
        time=data.get("time") or "",
        location=data.get("location") or "",
        club_id=club_id,
        status=event_status,
        created_by=created_by,
        attendance_count=data.get("attendanceCount"),
    )
    db.session.add(ev)
    db.session.commit()
    return jsonify(event_to_dict(ev)), 201


@bp.route("/<eid>", methods=["PUT"])
@jwt_required()
@role_required("admin", "club_head")
def update_event(eid):
    uid = get_jwt_identity()
    user = db.session.get(User, uid)
    if not user:
        return jsonify({"message": "Unauthorized"}), 401

    ev = db.session.get(Event, eid)
    if not ev:
        return jsonify({"message": "Not found"}), 404
    if not _can_manage_event(user, ev.club_id):
        return jsonify({"message": "Forbidden"}), 403

    if user.role == "club_head" and ev.status == "approved":
        return jsonify({"message": "Approved events cannot be modified by club heads"}), 403

    data = request.get_json(silent=True) or {}
    for field, key in [
        ("title", "title"),
        ("description", "description"),
        ("date", "date"),
        ("time", "time"),
        ("location", "location"),
    ]:
        if key in data:
            setattr(ev, field, data[key] or "")
    if "status" in data and user.role == "admin":
        ev.status = data["status"]

    if "attendanceCount" in data:
        ev.attendance_count = data["attendanceCount"]
    db.session.commit()
    return jsonify(event_to_dict(ev)), 200


@bp.route("/<eid>", methods=["DELETE"])
@jwt_required()
@role_required("admin", "club_head")
def delete_event(eid):
    uid = get_jwt_identity()
    user = db.session.get(User, uid)

    ev = db.session.get(Event, eid)
    if not ev:
        return jsonify({"message": "Not found"}), 404

    if not _can_manage_event(user, ev.club_id):
        return jsonify({"message": "Forbidden"}), 403

    db.session.delete(ev)
    db.session.commit()
    return "", 204


@bp.route("/<eid>/approve", methods=["POST"])
@jwt_required()
@role_required("admin")
def approve_event(eid):
    ev = db.session.get(Event, eid)
    if not ev:
        return jsonify({"message": "Not found"}), 404
    ev.status = "approved"
    db.session.commit()
    return jsonify(event_to_dict(ev)), 200


@bp.route("/<eid>/reject", methods=["POST"])
@jwt_required()
@role_required("admin")
def reject_event(eid):
    ev = db.session.get(Event, eid)
    if not ev:
        return jsonify({"message": "Not found"}), 404
    ev.status = "rejected"
    db.session.commit()
    return jsonify(event_to_dict(ev)), 200

File: backend/src/routes/events.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');

const { jwtOptional, jwtRequired, roleRequired } = require('../middleware/auth');
const User = require('../models/User');
const ClubMember = require('../models/ClubMember');
const Event = require('../models/Event');
const { eventToDict } = require('../serializers');

const router = express.Router();

function canManageEvent(user, clubId) {
  if (!user) return false;
  if (user.role === 'admin') return true;
  if (user.role === 'club_head' && user.clubId === clubId) return true;
  return false;
}

// Role-specific event display:
// - admin: all events
// - club_head: only their club's events
// - student: all events from clubs in which the student is enrolled + approved events
// - guest: approved events only (fallback)
router.get('/', jwtOptional, async (req, res) => {
  const uid = req.userId;

  if (!uid) {
    const events = await Event.find({ status: 'approved' });
    return res.status(200).json(await Promise.all(events.map(eventToDict)));
  }

  const user = await User.findById(uid);
  if (!user) {
    const events = await Event.find({ status: 'approved' });
    return res.status(200).json(await Promise.all(events.map(eventToDict)));
  }

  let events;
  if (user.role === 'admin') {
    events = await Event.find().sort({ date: -1 });
  } else if (user.role === 'club_head') {
    events = await Event.find({ clubId: user.clubId }).sort({ date: -1 });
  } else if (user.role === 'student') {
    const memberships = await ClubMember.find({ userId: uid });
    const clubIds = memberships.map((m) => m.clubId);
    events = await Event.find({
      $or: [{ status: 'approved' }, ...(clubIds.length ? [{ clubId: { $in: clubIds } }] : [])],
    }).sort({ date: -1 });
  } else {
    events = await Event.find({ status: 'approved' });
  }

  res.status(200).json(await Promise.all(events.map(eventToDict)));
});

router.get('/:eid', async (req, res) => {
  const ev = await Event.findById(req.params.eid);
  if (!ev) {
    return res.status(404).json({ message: 'Not found' });
  }
  res.status(200).json(await eventToDict(ev));
});

router.post('/', jwtRequired, roleRequired('admin', 'club_head'), async (req, res) => {
  const user = await User.findById(req.userId);
  if (!user) {
    return res.status(401).json({ message: 'Unauthorized' });
  }

  const data = req.body || {};
  const clubId = data.clubId || data.club_id;
  if (!clubId) {
    return res.status(400).json({ message: 'clubId required' });
  }
  if (!canManageEvent(user, clubId)) {
    return res.status(403).json({ message: 'Forbidden' });
  }

  const title = (data.title || '').trim();
  if (!title) {
    return res.status(400).json({ message: 'Title required' });
  }

  const eid = `event-${uuidv4().replace(/-/g, '').slice(0, 12)}`;
  const createdBy = data.createdBy || data.created_by || req.userId;
  const status = user.role === 'admin' ? 'approved' : 'pending';

  const ev = new Event({
    _id: eid,
    title,
    description: data.description || '',
    date: data.date || '',
    time: data.time || '',
    location: data.location || '',
    clubId,
    status,
    createdBy,
    attendanceCount: data.attendanceCount != null ? data.attendanceCount : null,
  });
  await ev.save();
  res.status(201).json(await eventToDict(ev));
});

router.put('/:eid', jwtRequired, roleRequired('admin', 'club_head'), async (req, res) => {
  const user = await User.findById(req.userId);
  if (!user) {
    return res.status(401).json({ message: 'Unauthorized' });
  }

  const ev = await Event.findById(req.params.eid);
  if (!ev) {
    return res.status(404).json({ message: 'Not found' });
  }
  if (!canManageEvent(user, ev.clubId)) {
    return res.status(403).json({ message: 'Forbidden' });
  }

  if (user.role === 'club_head' && ev.status === 'approved') {
    return res.status(403).json({ message: 'Approved events cannot be modified by club heads' });
  }

  const data = req.body || {};
  for (const key of ['title', 'description', 'date', 'time', 'location']) {
    if (key in data) {
      ev[key] = data[key] || '';
    }
  }
  if ('status' in data && user.role === 'admin') {
    ev.status = data.status;
  }
  if ('attendanceCount' in data) {
    ev.attendanceCount = data.attendanceCount;
  }
  await ev.save();
  res.status(200).json(await eventToDict(ev));
});

router.delete('/:eid', jwtRequired, roleRequired('admin', 'club_head'), async (req, res) => {
  const user = await User.findById(req.userId);

  const ev = await Event.findById(req.params.eid);
  if (!ev) {
    return res.status(404).json({ message: 'Not found' });
  }

  if (!canManageEvent(user, ev.clubId)) {
    return res.status(403).json({ message: 'Forbidden' });
  }

  await ev.deleteOne();
  res.status(204).send();
});

router.post('/:eid/approve', jwtRequired, roleRequired('admin'), async (req, res) => {
  const ev = await Event.findById(req.params.eid);
  if (!ev) {
    return res.status(404).json({ message: 'Not found' });
  }
  ev.status = 'approved';
  await ev.save();
  res.status(200).json(await eventToDict(ev));
});

router.post('/:eid/reject', jwtRequired, roleRequired('admin'), async (req, res) => {
  const ev = await Event.findById(req.params.eid);
  if (!ev) {
    return res.status(404).json({ message: 'Not found' });
  }
  ev.status = 'rejected';
  await ev.save();
  res.status(200).json(await eventToDict(ev));
});

module.exports = router;
